relays/finality-relay/src/finality_loop_tests.rs

```diff
@@ -0,0 +1,366 @@
+			(TestSourceHeader(false, 16), TestFinalityProof(Some(16))),
+		],
+	);
+}
+
+#[test]
+fn prune_unjustified_headers_works() {
+	let mut original_unjustified_headers: UnjustifiedHeaders<TestFinalitySyncPipeline> = vec![
+		TestSourceHeader(false, 10),
+		TestSourceHeader(false, 13),
+		TestSourceHeader(false, 15),
+		TestSourceHeader(false, 17),
+		TestSourceHeader(false, 19),
+	]
+	.into_iter()
+	.collect();
+
+	// when header is in the collection
+	let mut unjustified_headers = original_unjustified_headers.clone();
+	assert_eq!(
+		prune_unjustified_headers::<TestFinalitySyncPipeline>(10, &mut unjustified_headers),
+		Some(TestSourceHeader(false, 10)),
+	);
+	assert_eq!(
+		&original_unjustified_headers.clone().make_contiguous()[1..],
+		unjustified_headers.make_contiguous(),
+	);
+
+	// when there's no header in the collection
+	let mut unjustified_headers = original_unjustified_headers.clone();
+	assert_eq!(
+		prune_unjustified_headers::<TestFinalitySyncPipeline>(11, &mut unjustified_headers),
+		None,
+	);
+	assert_eq!(
+		&original_unjustified_headers.clone().make_contiguous()[1..],
+		unjustified_headers.make_contiguous(),
+	);
+
+	// when last entry is pruned
+	let mut unjustified_headers = original_unjustified_headers.clone();
+	assert_eq!(
+		prune_unjustified_headers::<TestFinalitySyncPipeline>(19, &mut unjustified_headers),
+		Some(TestSourceHeader(false, 19)),
+	);
+
+	assert_eq!(
+		&original_unjustified_headers.clone().make_contiguous()[5..],
+		unjustified_headers.make_contiguous(),
+	);
+
+	// when post-last entry is pruned
+	let mut unjustified_headers = original_unjustified_headers.clone();
+	assert_eq!(
+		prune_unjustified_headers::<TestFinalitySyncPipeline>(20, &mut unjustified_headers),
+		None,
+	);
+	assert_eq!(
+		&original_unjustified_headers.make_contiguous()[5..],
+		unjustified_headers.make_contiguous(),
+	);
+}
+
+#[test]
+fn prune_recent_finality_proofs_works() {
+	let mut original_recent_finality_proofs: FinalityProofs<TestFinalitySyncPipeline> = vec![
+		(10, TestFinalityProof(Some(10))),
+		(13, TestFinalityProof(Some(13))),
+		(15, TestFinalityProof(Some(15))),
+		(17, TestFinalityProof(Some(17))),
+		(19, TestFinalityProof(Some(19))),
+	]
+	.into_iter()
+	.collect();
+
+	// when there's proof for justified header in the deque
+	let mut recent_finality_proofs = original_recent_finality_proofs.clone();
+	prune_recent_finality_proofs::<TestFinalitySyncPipeline>(10, &mut recent_finality_proofs, 1024);
+	assert_eq!(
+		&original_recent_finality_proofs.clone().make_contiguous()[1..],
+		recent_finality_proofs.make_contiguous(),
+	);
+
+	// when there are no proof for justified header in the deque
+	let mut recent_finality_proofs = original_recent_finality_proofs.clone();
+	prune_recent_finality_proofs::<TestFinalitySyncPipeline>(11, &mut recent_finality_proofs, 1024);
+	assert_eq!(
+		&original_recent_finality_proofs.clone().make_contiguous()[1..],
+		recent_finality_proofs.make_contiguous(),
+	);
```

Comment:
It's not clear to me what happens here - does the first header (since its the most recent compared with the given header) get pruned?

Reply:
Yes